File: website/test/_lib/loopback.test.ts
import { describe, it, expect } from "vitest";
import { isLoopbackRedirectUri, safeUrl } from "../../functions/_lib/loopback";

describe("isLoopbackRedirectUri", () => {
  it("accepts http://127.0.0.1:<port>/... with any port", () => {
    expect(isLoopbackRedirectUri("http://127.0.0.1:8931/cb")).toBe(true);
    expect(isLoopbackRedirectUri("http://127.0.0.1:1/callback")).toBe(true);
    expect(isLoopbackRedirectUri("http://127.0.0.1:65535/x")).toBe(true);
  });

  it("rejects localhost hostname (DNS rebinding defense)", () => {
    expect(isLoopbackRedirectUri("http://localhost:8931/cb")).toBe(false);
  });

  it("rejects non-http scheme", () => {
    expect(isLoopbackRedirectUri("https://127.0.0.1:8931/cb")).toBe(false);
  });

  it("rejects non-loopback hosts", () => {
    expect(isLoopbackRedirectUri("http://192.168.1.5:8931/cb")).toBe(false);
    expect(isLoopbackRedirectUri("http://example.com/cb")).toBe(false);
    expect(isLoopbackRedirectUri("http://0.0.0.0:8931/cb")).toBe(false);
    // IPv6 loopback [::1] is intentionally not accepted by the 127.0.0.1 rule
    expect(isLoopbackRedirectUri("http://[::1]:8931/cb")).toBe(false);
  });

  it("rejects malformed / relative / empty URIs", () => {
    expect(isLoopbackRedirectUri("")).toBe(false);
    expect(isLoopbackRedirectUri("/cb")).toBe(false);
    expect(isLoopbackRedirectUri("127.0.0.1:8931/cb")).toBe(false);
    expect(isLoopbackRedirectUri("not a url")).toBe(false);
  });
});

describe("safeUrl", () => {
  it("returns URL for valid input and null for invalid", () => {
    expect(safeUrl("http://127.0.0.1:8931/cb")?.hostname).toBe("127.0.0.1");
    expect(safeUrl("garbage")).toBeNull();
    expect(safeUrl("")).toBeNull();
  });
});

File: server/internal/broker/loopback_test.go
package broker

import (
	"strings"
	"testing"
)

// TestIsLoopbackRedirectURI mirrors loopback.test.ts.
func TestIsLoopbackRedirectURI(t *testing.T) {
	accept := []string{
		"http://127.0.0.1:8931/cb",
		"http://127.0.0.1:1/callback",
		"http://127.0.0.1:65535/x",
	}
	for _, v := range accept {
		if !isLoopbackRedirectURI(v) {
			t.Errorf("isLoopbackRedirectURI(%q) = false, want true", v)
		}
	}

	reject := []string{
		"http://localhost:8931/cb",   // DNS-rebinding defense
		"https://127.0.0.1:8931/cb",  // non-http scheme
		"http://192.168.1.5:8931/cb", // non-loopback host
		"http://example.com/cb",
		"http://0.0.0.0:8931/cb",
		"http://[::1]:8931/cb", // IPv6 loopback not accepted by 127.0.0.1 rule
		"",
		"/cb",               // relative
		"127.0.0.1:8931/cb", // no scheme
		"not a url",
	}
	for _, v := range reject {
		if isLoopbackRedirectURI(v) {
			t.Errorf("isLoopbackRedirectURI(%q) = true, want false", v)
		}
	}
}

// TestIsLoopbackRedirectURILengthCap enforces decision 7's length cap.
func TestIsLoopbackRedirectURILengthCap(t *testing.T) {
	long := "http://127.0.0.1:8931/" + strings.Repeat("a", maxRedirectURILen)
	if isLoopbackRedirectURI(long) {
		t.Error("over-length loopback URI should be rejected")
	}
}

// TestSafeURL mirrors loopback.test.ts safeUrl.
func TestSafeURL(t *testing.T) {
	u := safeURL("http://127.0.0.1:8931/cb")
	if u == nil || u.Hostname() != "127.0.0.1" {
		t.Errorf("safeURL valid input = %v", u)
	}
	if safeURL("garbage") != nil {
		t.Error("safeURL(garbage) should be nil")
	}
	if safeURL("") != nil {
		t.Error("safeURL(empty) should be nil")
	}
}
